Minimal STL parser

MatWise opened the pull request:
Create a basic Siemens STL parser and integrate it into CBMC.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

tautschnig commented:
This looks really nice and clean overall - my only concern is the ambiguity of the acronym "STL." It just seems too easy to confuse with "Standard Template Library," especially when the framework also (partly) supports analysis of C++ code. Is there any chance to choose a different name?

MatWise committed:
Add basic Statement List parser

Includes support for simple arithmetic and boolean expressions as well as other
Statement List language features, such as networks, functions and function
blocks.

MatWise committed:
Integrate STL parser into CBMC executable

Includes a basic Statement List language interface (although without any
typechecking yet) as well as modifications to some Makefiles and
cbmc_languages.cpp. Use the --show-parse-tree option when parsing a .awl file
to see the project's output.

MatWise committed:
Add Statement List regression tests

Includes tests for simple integer additions/multiplications, multiple
functions/function blocks and for dividing two floats.

pkesseli commented on src/structured-text/converters/convert_int_value.cpp:
```diff
@@ -0,0 +1,32 @@
+  signed value = std::stoi(src);
+  typet type(ID_structured_text_int_value);
+
+  exprt result(std::to_string(value), type);
```
Why convert `src` to `int` and then back to `std::string`? Perhaps you can just:
```
return constant_exprt(src, typet(ID_stl_int_value));
```

pkesseli replied:
The `id` of an expression should be its kind, not its value. The `constant_exprt` constructor sets `id` to `ID_constant` and the `value` property to `src`. You will need to change `stl_parse_treet::output` to make your regression tests pass.

pkesseli commented on src/structured-text/converters/convert_int_value.cpp:
```diff
@@ -0,0 +1,32 @@
+exprt convert_int_value(const std::string &src)
+{
+  signed value = std::stoi(src);
+  typet type(ID_structured_text_int_value);
```
Use `signedbv_typet`:
```
  signedbv_typet type(STRUCTURED_TEXT_INT_WIDTH); // Appropriate width?
  type.set(ID_structured_text_type, ID_structured_text_int_value);
```

Perhaps create a helper similar to `c_types.h`.

pkesseli commented on src/structured-text/converters/convert_int_value.cpp:
```diff
@@ -0,0 +1,32 @@
+#include <util/std_expr.h>
+#include <util/std_types.h>
+
+exprt convert_int_value(const std::string &src)
```
Prefer `constant_exprt`.

pkesseli commented on src/structured-text/converters/convert_int_value.cpp:
```diff
@@ -0,0 +1,32 @@
+  return result;
+}
+
+exprt convert_bit_value(const std::string &src)
```
Same: `return constant_exprt(src, typet(ID_stl_bit_value));`

pkesseli commented on src/structured-text/converters/convert_int_value.h:
```diff
@@ -0,0 +1,24 @@
+#include <util/expr.h>
+
+/// Converts a string into the corresponding 'Int' expression.
+exprt convert_int_value(const std::string &src);
```
Document parameters and return values for functions:
```
/// ...
/// \param src ...
/// \return ...
```

pkesseli commented on src/structured-text/converters/convert_real_value.cpp:
```diff
@@ -0,0 +1,23 @@
+#include <util/std_expr.h>
+#include <util/std_types.h>
+
+exprt convert_real_value(const std::string &src)
```
See comments for `convert_int_value`.

pkesseli commented on src/structured-text/converters/convert_string_value.cpp:
```diff
@@ -0,0 +1,43 @@
+
+#include "convert_string_value.h"
+
+exprt convert_identifier(const std::string &src)
```
Both `convert_identifier` and `convert_title` can probably just return `string_constantt`.

pkesseli commented on src/structured-text/converters/convert_string_value.cpp:
```diff
@@ -0,0 +1,43 @@
+  return result;
+}
+
+exprt convert_label(const std::string &src)
```
Ideally we use a `code_labelt` here and eventually merge it with the subsequent `codet` - although that can happen at `typecheck`.

pkesseli commented on src/structured-text/converters/expr2structured_text.h:
```diff
@@ -0,0 +1,24 @@
+class namespacet;
+class typet;
+
+std::string expr2stl(const exprt &expr);
```
Prefer inline forward declarations:
```
std::string expr2stl(const class exprt &expr);
```

pkesseli replied:
Missing comments.

pkesseli commented on src/structured-text/structured_text_language.cpp:
```diff
@@ -0,0 +1,131 @@
+  exprt &expr,
+  const namespacet &ns)
+{
+  // unused parameters
```
Is this to avoid compiler warnings?

MatWise replied:
Yes, although I don't know if this is necessary anymore (`ansi_c_languaget` does it in a similar way at least).

romainbrenguier replied:
Would it work to not name them instead (`to_expr(const std::string &, const std::string &, ...)`)?

pkesseli commented on src/structured-text/structured_text_language.h:
```diff
@@ -0,0 +1,140 @@
+    exprt &expr,
+    const namespacet &ns) override;
+
+  std::unique_ptr<languaget> new_language() override
```
Move implementations whenever possible to `.cpp`.

pkesseli commented on src/structured-text/structured_text_parser.cpp:
```diff
@@ -0,0 +1,269 @@
+  // Generate new function block.
+  structured_text_parse_treet::function_blockt fb(
+    current_name, current_version);
+  current_parsed_root_element = &fb;
```
With very few exceptions, raw pointers are not allowed in the CBMC code base. Just pass `fb` as an argument by reference to `iterate_variables` and `iterate_networks` instead.

pkesseli commented on src/structured-text/structured_text_parser.cpp:
```diff
@@ -0,0 +1,269 @@
+  current_parsed_root_element = &fb;
+
+  // Fill the block with networks and variables.
+  iterate_variables();
```
Perhaps `fill_variables` and `fill_networks` would be a better name choice?

pkesseli commented on src/structured-text/structured_text_parser.cpp:
```diff
@@ -0,0 +1,269 @@
+    function.id() == ID_structured_text_function,
+    "Expression ID should be stl_function");
+
+  current_root_element = function;
```
Don't use member variables unless necessary. Just pass `function` to the functions that need access to it.

pkesseli commented on src/structured-text/structured_text_parser.cpp:
```diff
@@ -0,0 +1,269 @@
+  current_root_element = function;
+
+  // Search for the correct tree elements and set the internal parser state.
+  INVARIANT(find_name(), "Function should have a name");
```
Since the whole function is created in the scope of the invocation of `add_function`, there is no need to work with members. Just have `find_version` etc. return their values, and move the invariant inside those functions.

pkesseli commented on src/structured-text/structured_text_parser.h:
```diff
@@ -0,0 +1,95 @@
+  structured_text_parse_treet::functiont *current_parsed_root_element;
+  structured_text_parse_treet::networkt current_parsed_network;
+
+  bool find_name();
```
Document private members as well.

pkesseli commented on src/structured-text/structured_text_typecheck.cpp:
```diff
@@ -0,0 +1,53 @@
+void structured_text_typecheckt::typecheck()
+{
+  for(const auto &fn : parse_tree.functions)
+  {
```
Omit `{` and `}` for single-line bodies.

romainbrenguier commented on src/structured-text/converters/convert_int_value.cpp:
```diff
@@ -0,0 +1,27 @@
+constant_exprt convert_int_value(const std::string &src)
+{
+  signedbv_typet type(16);
+  type.set(ID_structured_text_type, ID_structured_text_int_value);
```
It looks like there should be a `structured_text_int_value_typet` or similar instead of reusing `signedbv_typet` and adding a flag

pkesseli replied:
Integers in STL are bitvectors, so they should be modelled this way. Both the Java and ANSI C parser use `signedbv_typet` in this fashion, e.g.:
```
signedbv_typet signed_int_type()
{
  signedbv_typet result(config.ansi_c.int_width);
  result.set(ID_C_c_type, ID_signed_int);
  return result;
}
```

However, I suggest creating an `stl_types.h` equivalent to `c_types.h` with:
```
#define STL_INT_WIDTH 16u

signedbv_typet stl_int_type()
{
  signedbv_typet result(STL_INT_WIDTH); 
  result.set(ID_structured_text_type, ID_structured_text_int_value); 
  return result;
}
```

And perhaps renaming `ID_structured_text_int_value` to just `ID_structured_text_int`.

MatWise replied:
Renamed the IDs accordingly, I'd add `stl_types.h` in a future PR.

romainbrenguier commented on src/structured-text/converters/convert_int_value.cpp:
```diff
@@ -0,0 +1,27 @@
+
+constant_exprt convert_int_value(const std::string &src)
+{
+  signedbv_typet type(16);
```
:pick: brace initialisation is preferred

romainbrenguier commented on src/structured-text/converters/convert_int_value.cpp:
```diff
@@ -0,0 +1,27 @@
+{
+  signedbv_typet type(16);
+  type.set(ID_structured_text_type, ID_structured_text_int_value);
+  return constant_exprt(src, type);
```
this seems very error prone as the irep_idt in `constant_exprt` can be interpreted in a different way than the string is (in particular the base may not correspond), I would have expected some conversion from `src` to int to happen and then `from_integer` to be used.

pkesseli replied:
STL integer literal syntax doesn't match the one of C. So far, only `123`, `-123` and `+123` appear to be allowed for `INT` types in the STL grammar. I'd suggest at least removing stray `+` characters, which should make them compatible without explicit `std::stoi` call.

However, perhaps add a `TODO` to later introduce regression tests for invalid inputs, in order to check that the grammar properly rejects incorrectly specified literals.

MatWise replied:
Added functions for different Int literal representations. I'll add the regression tests once the assignment of literals is supported by the grammar.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 138102e85c66b0ecc84842d234c0a68d65016fcf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113086645
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj commented on unit/CMakeLists.txt:
```diff
@@ -57,6 +57,7 @@ target_link_libraries(
         goto-symex
         cbmc-lib
         json-symtab-language
+	structured-text
```
:pick: Incorrect indent

allredj commented on unit/Makefile:
```diff
@@ -155,6 +155,7 @@ CPROVER_LIBS =../src/ansi-c/ansi-c$(LIBEXT) \
               ../src/assembler/assembler$(LIBEXT) \
               ../src/analyses/analyses$(LIBEXT) \
               ../src/solvers/solvers$(LIBEXT) \
+	      ../src/structured-text/structured-text$(LIBEXT) \
```
:pick: Incorrect indent

allredj commented:
This compiles with TG fine. You can disregard the bot message (@MatWise is a "first time contributor" and TG CI does not fetch his fork). :heavy_check_mark:

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f4cf42155fc9e03b9f910368782fe5ae781b1ec2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113109006
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

tautschnig commented on src/structured-text/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+
+generic_includes(structured-text)
+
+target_link_libraries(structured-text util)
```
Nit pick: GitHub complains about a missing end-of-line

tautschnig commented on src/structured-text/converters/convert_int_literal.h:
```diff
@@ -0,0 +1,35 @@
+
+#include <string>
+
+#include <util/expr.h>
```
Nit pick: unnecessary include (only the next one is needed)

pkesseli commented on src/structured-text/converters/convert_bool_literal.cpp:
```diff
@@ -0,0 +1,26 @@
+{
+  // Transform to lower case.
+  std::string copy;
+  std::transform(src.begin(), src.end(), copy.begin(), ::tolower);
```
This is undefined behaviour. Replace by:
```
std::string copy(src.size());
transform(begin(src), end(src), begin(copy), ::tolower);
```

pkesseli replied:
Note: argument-dependent lookup (ADL) should allow you to just specify `transform` instead of `std::transform`.

pkesseli replied:
Prefer `begin(src)` over `src.begin()`, etc.

pkesseli commented on src/structured-text/converters/convert_bool_literal.cpp:
```diff
@@ -0,0 +1,26 @@
+
+constant_exprt convert_bool_literal(const std::string &src)
+{
+  // Transform to lower case.
```
I think the code is clear enough by itself. Remove comment.

pkesseli commented on src/structured-text/converters/convert_int_literal.cpp:
```diff
@@ -0,0 +1,50 @@
+
+constant_exprt convert_int_dec_literal(const std::string &src)
+{
+  // Remove '+' char if present.
```
Code is clear enough by itself. Avoid comment clutter.

pkesseli commented on src/structured-text/converters/convert_int_literal.cpp:
```diff
@@ -0,0 +1,50 @@
+constant_exprt convert_int_hex_literal(const std::string &src)
+{
+  // Cut the leading '16#' and convert to int
+  std::string cut = src.substr(3, std::string::npos);
```
Extract `magic numbers` to `#define` constants, document those and remove the above comment.

pkesseli replied:
const std::string cut{src.substr(...)};

pkesseli replied:
Use expressive names, i.e. `const std::string hex_literal = ...`

tautschnig commented on src/structured-text/converters/convert_string_value.h:
```diff
@@ -0,0 +1,34 @@
+#ifndef CPROVER_STRUCTURED_TEXT_CONVERTERS_CONVERT_STRING_VALUE_H
+#define CPROVER_STRUCTURED_TEXT_CONVERTERS_CONVERT_STRING_VALUE_H
+
+#include <util/expr.h>
```
Seems unnecessary

tautschnig commented on src/structured-text/converters/expr2structured_text.cpp:
```diff
@@ -0,0 +1,28 @@
+std::string expr2stl(const exprt &expr)
+{
+  // TODO: Implement.
+  return "";
```
How about using `return expr.pretty();` for a slightly more useful approximation?

tautschnig commented on src/structured-text/parser.y:
```diff
@@ -0,0 +1,701 @@
+/*** STL file structure keywords *********************************************/
+%token TOK_VERSION 		   		"VERSION"
+%token TOK_BEGIN   		   		"BEGIN"
+%token TOK_FUNCTION_BLOCK  		"FUNCTION_BLOCK"
```
Formatting looks weird here - mix of tabs and spaces? (Please don't use tabs.)

pkesseli commented on src/structured-text/converters/convert_int_literal.cpp:
```diff
@@ -0,0 +1,50 @@
+{
+  // Remove '+' char if present.
+  std::string value(src);
+  value.erase(std::remove(value.begin(), value.end(), '+'), value.end());
```
ADL:
`value.erase(remove(begin(value), end(value), '+'), end(value));`

tautschnig commented on src/structured-text/parser.y:
```diff
@@ -0,0 +1,701 @@
+	init FB_Decl
+	| init Func_Decl
+	| /* nothing */
+	;
```
Nit picking: most rules appear to use 4-character indent, but a few others use 8 characters or so?!

tautschnig commented on src/structured-text/scanner.l:
```diff
@@ -0,0 +1,191 @@
+[\t\r\n ]			;
+:=					{ loc(); return TOK_ASSIGNMENT; }
+BEGIN				{ loc(); return TOK_BEGIN; }
+VERSION				{ loc(); return TOK_VERSION; }
```
The indentation is a bit strange.

tautschnig commented on src/structured-text/structured_text_language.cpp:
```diff
@@ -0,0 +1,168 @@
+
+void structured_text_languaget::set_language_options(const optionst &options)
+{
+  (void)options;
```
Just remove the names of unused parameters.

tautschnig commented on src/structured-text/structured_text_language.h:
```diff
@@ -0,0 +1,115 @@
+{
+public:
+  /// Sets language specific options.
+  void set_language_options(const optionst &options) override;
```
No need to include parameter names when the documentation doesn't refer to them.

tautschnig commented on src/structured-text/structured_text_parser.cpp:
```diff
@@ -0,0 +1,259 @@
+      return op;
+  }
+  INVARIANT(false, "Root expression should have a variable list");
+  return exprt(ID_nil);
```
Use `nil_exprt{}` (as that uses a single, shared irept). Though you could actually also use `UNREACHABLE` and then not add any `return` statement.

MatWise replied:
I tried that first, but Eclipse (cpplint?) complains about it. Is there any way to suppress warnings like that properly?

tautschnig replied:
Hmm, that's surprising as we already use `UNREACHABLE` in that way in several places - or maybe Eclipse also warns in those cases?

tautschnig reviewed (approved):
This is overall very nicely done!!! Just one request in addition to all the nit-picks below: could you please make sure that all "TODO" comments include sufficient information for someone else to pick this up and work on it?

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f3b4b65213fb50d4aabe1863179a68803a435b15).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113740279
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: eb31831253a56d7e11fd3bff0ea86859d38bbdd5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113772968
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d484911e0f401bc932e1aeb33182aa2ff349ace2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113899257
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d666742d88c7df262beb55885dbe052ca51dd028).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113901720
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2ba1361886efb5d6949e453440ee2e8b26c8c4fc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113911945
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
